include/pressio/type_traits/have_matching_exe_space.hpp

```diff
@@ -49,24 +49,36 @@
 #ifndef TYPE_TRAITS_HAVE_MATCHING_EXE_SPACE_HPP_
 #define TYPE_TRAITS_HAVE_MATCHING_EXE_SPACE_HPP_
 
-namespace pressio{ 
+namespace pressio{
+namespace impl {
+
+template<typename T1, typename T2, typename Enable = void>
+constexpr bool have_same_execution_space_v = false;
+
+template<typename T1, typename T2>
+constexpr bool have_same_execution_space_v<
+  T1, T2,
+  mpl::enable_if_t<
+       ::pressio::impl::has_execution_space_v<T1>
+    && ::pressio::impl::has_execution_space_v<T2>
+    >
+  > = std::is_same<
+    typename ::pressio::impl::execution_space<T1>::type,
+    typename ::pressio::impl::execution_space<T2>::type
+    >::value;
+
+} // namespace impl
 
 template <typename ... Args>
-struct have_matching_execution_space;
+struct have_matching_execution_space : public std::false_type {};
```

Comment:
this seems weird to me, how can this not conflict with the one below T1, T2, ...rest? 
Also, I don't think this ever get instantiated. I would remove the : public std::false_type{}. Not needed.